crates/codegen/parser/runtime/src/support/parser_function.rs

```diff
@@ -17,11 +18,35 @@ where
 
 impl<L, F> ParserFunction<L> for F
```

Comment:
> return a "tree" that has both parents - the first one being the expected parsed rule node but also the second parent being the "leading trivia"

That would be my preference as well, matching #640 .. IMO it is the cleanest solution without introducing other hacks/workarounds like virtual `EndOfFile` tokens, and lets the user analyzer exactly only the result they requested to parse, allowing checking trivia separately via a separate property, e.g. `ParseOutput::trailing_trivia`

Reply:
I'm _somewhat_ torn on having two separate channels for the parse output (one being the main tree, the other being the end-of-file trivia tacked on in the `ParseOutput`).

Is it a reasonable guarantee to always return a single "root" node for the parse output or can we reasonably output multiple top-level nodes? I know we rely on this internally but I wonder if that's also what the user expects.

@OmarTawfik  @AntonyBlakey thoughts?

Reply:
I don't know about how we "rely on this internally", I think you are more familiar with that logic. But as far as user expectations, when parsing any individual language element, for example `Expression`, I think it is reasonable to expect only one `Expression` node at the root, and nothing else. They shouldn't try to parse an expression, and then have to post-process the output to extract that expression.

This leads to the question: What should we do about an input like below?

```solidity
x + y
// comment
```

We have three solutions:

- Add a new `EndOfInput` virtual token/TokenKind, to attach the comment to. But we are really spending a lot of effort to clean up such virtual/magical kinds, and I don't want to add one more here. It is unexpected. It is not part of the user-written language definition. It is not structurally/logically part of the statement. And more importantly, since we are planning on promoting trivia to be siblings to other tokens, it won't really serve any other purpose, as it will just be another sibling to include alongside other trivia under the top parent. We can just insert the trivia without it.
- Consider it a parse error, because the input is no longer just an expression, and error on the `//comment` being unexpected/extra input. But it feels strange for a comment to produce syntax errors, since the majority of users (and a good number of tools/parsers/linters) completely skip them. I think it is a worse UX for users.
- Collect the comment into the `ParseOutput::trailing_trivia` field. It seems the most varstile option, as the input is parsed correctly, and if the user doesn't care about trivia, they never have to think twice about it. But if they do, they can easily access it separately, without polluting the main CST.

I think the same reasoning applies to the top-level `SourceUnit`, although it is a bit more common to see comments at the end of the file.

I wonder what is your worry about the new separate channel?

Reply:
> I don't know about how we "rely on this internally" (...)

`ParseOutput::parse_tree` is a single `Node`; we also assert that there is only a single top-most node after parsing, inside `ParserFunction::parse` here.

> I wonder what is your worry about the new separate channel?

We now have two places we need to inspect to see what has been parsed and now the caller has to stitch the output themselves to have a single parse tree.

Consider this example (added a CST test in https://github.com/NomicFoundation/slang/pull/733/commits/738692141ef3aba6b5b8c0c63823863dd7bb4a2c):
```solidity
contract X { }

/* new lines after this */

unexpectedchars
```

If we collect the "trailing leading"/"end of file" trivia into a separate place, without including it in the tree, we now end up with a hole in the tree:
```yaml
tree:
  - ContractDefinition: # 0..64?
    - ActualNodes: ... # 0..15
    # Hole: this is where the trivia is collected from
    - SKIPPED: 'unexpectedchars' # 48..64

trailing_trivia: # 15..48
  - MultilineComment: ...
```

> we are planning on promoting trivia to be siblings to other tokens (...)

I know that we briefly talked about it as part of the move to DSL v2 but I feel it was lumped together with a lot of different things and I don't recall having a strong consensus on this and what this might imply, especially now that we have named nodes and having multiple children nodes with the same name is a temporary property rather than explicit design choice (e.g. it might affect CST query design; @AntonyBlakey do we need to revisit that decision soon?).

Can we brainstorm this again, ideally also in the context of how might approach parsing NatSpec? I feel like this will or might strongly influence on how we approach (structured) trivia and I'd avoid commiting to superficial but requiring extra work changes like this if there's a risk of redesigning it again soon.

To not introduce a token, I think the current approach of collecting "lone" LeadingTrivia at the end is the most versatile, as it effectively replaces EndOfFileTrivia in this regard.

We can (and probably will have to, eventually) introduce `ParseOptions` that can have a setting of `collect_remaining_trivia`, defaulting to `true`. This way, the user is informed through documentation/API of different approaches the parser can take (none of them more correct, in a way) and if they want to be super strict or ensure no extra trivia after, they can turn this off (much like we should, for the top-level trivia parse).

Reply:
> it might affect CST query design; @AntonyBlakey do we need to revisit that decision soon

It merely makes the queries more complicated/less clear. I would prefer to not have duplicates *unless* they represent repetition.

> Can we brainstorm this again, ideally also in the context of how might approach parsing NatSpec? I feel like this will or might strongly influence on how we approach (structured) trivia

I think we shouldn't talk of 'structured trivia', as if it were a special case of trivia processing. Rather it is an instance of general embedding.

Reply:
And IMO these decisions should always be measured against Simple vs. Easy, especially when UX is mentioned. Sometimes UX seems to be shorthand for 'convenient' as a synonym or 'easy'. Our approach to trivia should IMO be dominated by regularity and clarity with convenience provided by optional layered mechanisms.

Not suggesting anything is wrong here, just something triggered me into restating my oft-stated opinion :)

Reply:
> but I feel it was lumped together with a lot of different things and I don't recall having a strong consensus on this and what this might imply, especially now that we have named nodes and having multiple children nodes with the same name is a temporary property rather than explicit design choice

Happy to revisit this, but it might make sense to do that discussion f2f, as it touches a few other long-term changes. WDYT?

Reply:
Also, thinking more about your example above about the hole, wouldn't the comment be `LeadingTrivia` for the `SKIPPED` token in that case?

```
tree:
  - ContractDefinition: # 0..64?
    - ActualNodes: ... # 0..15
    - LeadingTrivia:
        - MultilineComment: ...
    - SKIPPED: 'unexpectedchars' # 48..64

trailing_trivia: [] # nothing!
  ```

Reply:
The whole premise, from what I understand, was to get rid of the virtual tokens (like leading trivia on EOF token), so I'm not sure how this changes anything :sweat_smile: We'd now have leading trivia on a SKIPPED token, that we initially agreed to replace with a new `Node` variant.

Reply:
If we're fine with the example above, I'd say we should just bite the bullet and introduce an EOF token for this case and/or for the case of empty input (but then, maybe a hypothetical `Node::Missing` would fit better?).

Reply:
I think we would have to have `Node::Missing` and `Node::Unrecognized` (what we call `SKIPPED` now) at some point, but not blocking to this PR 👍

Reply:
So, discussed f2f, we would eventually remove all the leading/trailing rule kinds, and promote trivia as siblings.